### PR TITLE
Rewrite bottle automation

### PR DESCRIPTION
**Pull request by benesch**
This adds a script that automates the whole process and improves the documentation around the process.

### CODE REVIEW

**Comment by benesch** [comment]
@quodlibetor please take a look at this when you get the chance. It absorbs much of the automation from #14, and adds some additional scripting to automatically update the formula appropriately.

On a meta note, I'm nervous about the amount of complexity involved in this. More stuff that can go wrong; more stuff that people don't understand. The old manual checklist did have the desirable property that it was more clear what was going on. But if we're going to automate, I'd rather go all the way and remove humans from the equation entirely.

File: bin/bump-version [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+echo "Updating formula for new source tarball..."
+sed -f /dev/stdin Formula/materialized.rb > Formula/materialized.rb.new <<EOF
+/^  url/c\\
+\  url "$url"
+/^  sha256/c\\
+\  sha256 "$archive_sha"
+/^  STABLE_BUILD_SHA/c\\
+\  STABLE_BUILD_SHA = "$build_sha".freeze
+EOF
+mv Formula/materialized.rb{.new,}
```
Comment (quodlibetor) [inline bin/bump-version:34]: would it be more straightforward to just have a template file that we sed into place?

doing `sed -e "s/MZ_VERSION/$mz_version/"  -e "s/MZSTABLE_BUILD_SHA/$stable_sha/" .. Build.template.rb > Formula/materialized.rb` seems less fragile and easier to debug (since you can just grep for the strings).

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -20,9 +20,10 @@ jobs:
         set -euo pipefail
         mkdir -p "$(dirname $(brew --repo ${{github.repository}}))"
         cp -a "$GITHUB_WORKSPACE" "$(brew --repo ${{github.repository}})"
-        brew install gnu-tar
-        brew install gzip
-        hash -r gzip
+        brew install docker docker-machine
+        brew services start docker-machine
+        docker-machine create --driver virtualbox --virtualbox-boot2docker-url=https://github.com/boot2docker/boot2docker/releases/download/v19.03.5/boot2docker.iso default
+        eval $(docker-machine env default)
```
Comment (quodlibetor) [inline .github/workflows/main.yml:26]: is there a reason this needs to be docker-machine instead of the Docker cask?

File: bin/bump-version [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+set -euo pipefail
+
+if [[ $# -ne 1 ]]; then
+  echo "usage: $0 VERSION" >&2
+  exit 1
+fi
+
+cd "$(dirname "$0")/.."
+
```
Comment (quodlibetor) [inline bin/bump-version:13]: could we check that `~/.netrc` exists and is configured at least vaguely correctly (`grep bintray $HOME/.netrc"`) here?

Comment (quodlibetor) [reply bin/bump-version:13]: (obviously only if we're not in CI)

File: bin/mkbottle [hunk 4]
```diff
@@ -23,76 +10,29 @@ trap "rm -rf $tmpdir" EXIT
...
+
+version=${1#v}
+
+container=$(docker run --rm -d buildpack-deps:focal-curl sleep 999999)
+trap "docker kill $container > /dev/null" EXIT
+
+docker cp Formula/materialized.rb "$container:/materialized.rb"
+docker cp misc/mkbottle-inner.bash "$container:/mkbottle.bash"
+docker exec -i "$container" bash /mkbottle.bash "$version"
+docker cp "$container:/bottle.tar.gz" bottle.tar.gz
```
Comment (quodlibetor) [inline bin/mkbottle:38]: Would just mounting a directory be simpler?

**Comment by quodlibetor** [comment]
Meta response:

I agree strongly with removing humans from the equation entirely, this brings us close, I think, to running this script in CI when we push a tag to the main materialize repo, which I think needs to be the goal?

I agree that we want to avoid complexity, but I think a lot of the complexity here comes from the verification of the fact that what we're uploading is in fact the same thing as it was tagged as? Perhaps we can avoid that check (and therefore lots of the complexity) by only publishing new bottles from CI? I believe that would be exactly as safe as what we're doing?

**Comment by benesch** [comment]
Yeah, life would be great if we could treat this repository as a deployment artifact of the main MaterializeInc/materialize, but there are a number of structural problems in the way that I don't see getting resolved anytime soon.

The first is that doing any sort of testing with Homebrew requires some form of containerization, to stay sane. Our macOS setup in Buildkite is... not that; it's about as manual as it gets. (I emailed someone from MacStadium, and six hours later they set us up with a dedicated macOS server on which I manually installed Rustup.) Running Cargo without a container is more or less ok, but running these brew test commands is a guaranteed way to totally corrupt the state of that agent.

The cheapest/fastest/(only?) way I know of to get proper macOS containers is to use a hosted CI provider like Travis or, as of recently, GitHub Actions. Unfortunately the hardware is _tiny_, and it legit takes 25m+ to build materialized in release mode on GitHub actions. So we can't use GitHub Actions for all of our macOS CI needs. We'd end up in a massive deploy job backup during working hours, and possibly permanently if the eng team gets much larger.

So we end up in the situation that we have: Buildkite does the hard work of building a release binary, and we perform the Homebrew tests in GitHub Actions here. 

Obnoxiously, we can't have Buildkite publish the bottles directly, because the contents of the bottle depend on the formula in this repository, and you end up with a circular dependency. Plus if we ever add a new runtime dependency, that needs to result in a change to the Homebrew formula. Either we need to test the bottle on every commit in MaterializeInc/materialize, or we need some sort of manual step here where a human can edit the formula to account for the new dependency.

It _is_ possible for CI here to be responsible for building and pushing the bottles to Bintray—this is what the main Homebrew repository does—but you have another miserable circular dependency that results in as much complexity as the reproducible builds, just in a different place. Like, if it's not _me_ whose responsible for uploading the bottles to Bintray, but CI, then I need to open a version bump PR without the bottles... and then I'm not even sure what. Maybe CI would update that same PR after it built the bottles? Or we'd have to do what Homebrew does (see e.g. `brew publish-pr`), where maintainers run `brew publish-pr`, which auto-updates the bottle stanza, instead of pressing the big green merge button.

The tl;dr is that the self-referential nature of Homebrew bottle stanzas makes for some insanely complicated CI pipelines. This is exactly the problem that mzbuild solved, now that I think about it :D. Ideally you'd find the bottle for a formula by asking "does bintray.com/homebrew/bottles/FORMULA/VERSION" exist, without embedding the SHA of the bottle in the formula.

_Very_ down to find a clean sane solution to all this, but I've been down a lot of the obvious roads, and they all lead to some form of madness.

File: .github/workflows/main.yml (hunk 2 above)
Comment (benesch) [reply .github/workflows/main.yml:26]: There is no way to non-interactively give Docker.app the permissions it needs to start a VM sadly. I tried pretty hard and couldn't make anything work. But since these agents already have virtualbox installed and running, using docker-machine is much cheaper than you might expect.

File: bin/mkbottle (hunk 4 above)
Comment (benesch) [reply bin/mkbottle:38]: So much simpler, and yet Docker for Mac refuses by default to bind mount volumes in `/usr/local`, plus you have to be super super careful to not accidentally create root-owned files. Doing it this way is ugly, but it localizes the ugly to the container.

File: bin/bump-version (hunk 1 above)
Comment (benesch) [reply bin/bump-version:34]: Ah, maybe, but there are some complexities to work out there too because you need to render _twice_, once without the bottle stanza and once with. Going to merge as is for now, but if you figure out how to make it cleaner I'm all for it.

File: bin/bump-version (hunk 3 above)
Comment (benesch) [reply bin/bump-version:13]: 👍 the beauty of the bump-version script is that it never runs in CI

Comment (benesch) [reply bin/bump-version:13]: (going to address in a follow-up)
